File: Assignment4/bankers.c
#include <stdio.h>

int main() {
    int n, m, i, j, k, flag = 0;
    printf("Enter number of processes: ");
    scanf("%d", &n);
    printf("Enter number of resources: ");
    scanf("%d", &m);

    int need[n][m], max[n][m], allocation[n][m], available[m], work[m];
    printf("Enter allocation matrix\n");
    for (i = 0; i < n; i++) {
        for (j = 0; j < m; j++) {
            scanf("%d", &allocation[i][j]);
        }
    }
    printf("Enter max matrix\n");
    for (i = 0; i < n; i++) {
        for (j = 0; j < m; j++) {
            scanf("%d", &max[i][j]);
        }
    }
    printf("Enter available matrix\n");
    for (j = 0; j < m; j++) {
        scanf("%d", &available[j]);
    }

    // Initializing need matrix
    for (i = 0; i < n; i++) {
        for (j = 0; j < m; j++) {
            need[i][j] = max[i][j] - allocation[i][j];
        }
    }

    // Initializing work array
    for (j = 0; j < m; j++) {
        work[j] = available[j];
    }

    // Check for safe state
    for (i = 0; i < n; i++) {
        for (j = 0; j < m; j++) {
            if (need[i][j] > work[j]) {
                flag = 1;
                break;
            }
        }
    }
    if (flag == 1) {
        printf("Unsafe State\n");
        return 0;
    }

    int finish[n];
    for (i = 0; i < n; i++) {
        finish[i] = 0;
    }

    int count = 0;
    int safeSeq[n];
    while (count < n) {
        flag = 0;
        for (i = 0; i < n; i++) {
            if (finish[i] == 0) {
                for (j = 0; j < m; j++) {
                    if (need[i][j] > work[j]) {
                        flag = 1;
                        break;
                    }
                }
                if (flag == 0) {
                    for (k = 0; k < m; k++) {
                        work[k] = work[k] + allocation[i][k];
                    }
                    safeSeq[count] = i;
                    finish[i] = 1;
                    count++;
                }
            }
        }
    }

    printf("Safe sequence: ");
    for (i = 0; i < n; i++) {
        printf("%d ", safeSeq[i]);
    }
    printf("\n");
    return 0;
}

File: Assignment4/bankers.py
# A Python3 program to implement the Banker's Algorithm
 
# Function to find the need of each process
def calculate_need(need, maxm, allot):
    for i in range(len(need)):
        for j in range(len(need[0])):
            need[i][j] = maxm[i][j] - allot[i][j]
 
# Function to find the system is in safe state or not
def is_safe(processes, avail, maxm, allot):
    need = [[0] * len(avail) for _ in range(len(processes))]
    calculate_need(need, maxm, allot)
    finished = [0] * len(processes)
    safe_sequence = []
    work = [a for a in avail]
 
    while 0 in finished:
        found = False
        for i in range(len(processes)):
            if finished[i] == 0:
                for j in range(len(avail)):
                    if need[i][j] > work[j]:
                        break
                else:
                    found = True
                    for j in range(len(avail)):
                        work[j] += allot[i][j]
                    finished[i] = 1
                    safe_sequence.append(processes[i])
        if not found:
            return (False, [])
 
    return (True, safe_sequence)
 
# Driver code
if __name__ == "__main__":
    # Total number of processes
    n_processes = 5
    # Total number of resources
    n_resources = 3
 
    # Available resources
    available = [3, 3, 2]
 
    # Maximum resources that can be allocated to each process
    maximum = [[7, 5, 3], [3, 2, 2], [9, 0, 2], [2, 2, 2], [4, 3, 3]]
 
    # Resources that are currently allocated to each process
    allocated = [[0, 1, 0], [2, 0, 0], [3, 0, 2], [2, 1, 1], [0, 0, 2]]
 
    # Check if the system is in safe state or not
    is_safe, safe_sequence = is_safe(list(range(n_processes)), available, maximum, allocated)
    if is_safe:
        print("The system is in safe state")
        print("Safe sequence is:", safe_sequence)
    else:
        print("The system is in unsafe state")
